chronos/apps/http/httpserver.nim

```diff
@@ -506,15 +521,55 @@ proc preferredContentType*(acceptHeader: string,
         # If `Accept` header is incorrect, client accepts any type of content.
         ok(types[0])
       else:
-        selectContentType(ares.get().data, types)
+        # This algorithm exploits ``q-value`` range which is [0.000, 1.000],
+        # by using smaller values for sorting server's preferred content types.
+        # So between 0.000 and 0.001 there present 99 possible values
+        # [0.00001, 0.00099] which can be used.
+        # Server's preferred types marked with `preferred-value` which is
+        # has highest value for first index and the lowest for the last.
+
+        # ``maxWeight`` represents maximum possible weight value which can be
+        # obtained.
+        let maxWeight = 1.0 + float(len(types)) / float(100_000)
+        var
+          currentType = MediaType()
+          currentWeight = 0.0
+        for itemType in ares.get().data:
+          let
+            preferredIndex = types.find(itemType.mediaType)
+            preferredWeight =
+              if preferredIndex == -1:
+                0.0
+              else:
+                # Calculate weight which depends on position in ``types``
+                # array. Because ``preferredIndex`` is always less than
+                # ``len(types)`` this weight could not be ``0.0``.
+                float(len(types) - preferredIndex) / float(100_000)
```

Comment:
These float divisions seems unnecessary to me. You can just make the comparison below that checks for a new best element to work as if it compares tuples: `(qvalue, preferredIndex)`. (you can either use a tuple directly, or you can code the logical equivalent of this tuple comparison).

Reply:
Clients must not violate the spec and if client will supply something like `q=0.0001` it will be rejected at earlier phase.